### ISSUE TITLE
Mapping all exceptions to error_codes

### ISSUE THREAD

**Issue by ChrisBFX** [randombit/botan#1742]
Somewhat related to #1299
I'm currently trying to map all exceptions from Botan to boost::system::error_codes and it somewhat works (see https://github.com/krumedia/asio-botan-stream/blob/master/include/asio/botan/detail/ConvertExceptions.h).
The hundreds of throws inside Botan, which don't use special exception classes but only have a error message string are quite annoying though. These make it quite hard to debug cause i can only map those to "unknown error".

Would it be feasible to change those throws to include a distinct error-code from an enum? This should also include the TLS_Exceptions, cause the alert type is often not enough.

**Comment by randombit** [comment]
> Would it be feasible to change those throws to include a distinct error-code from an enum? 

Do you mean a distinct error code attached to each and every exception? That seems kind of excessive to me.

But, some changes related to that that do seem reasonable:

- Never throw bare `Exception` (make the constructors protected) but always some descriptive subclass, since it is true, catching a base `Exception` just tells you "something went wrong" with no real hint unless you parse the message. (And, specific content of exception messages are not considered part of the stable API, so that doesn't really work anyway.)
- Have a reason code enum ala
```
enum class ErrorType {
  InvalidArgument,
  InternalError,
  ...
}
```
exposed by new virtual `Exception::error_type() const nothrow` which kind of reflects the existing structure of Exception types. That would also simplify code like https://github.com/randombit/botan/blob/master/src/lib/ffi/ffi.cpp#L38 which is doing something much like what you linked to.
- Capture `__FILE__`, `__LINE__`, and `__func__` in exceptions (probably with the help of throwing helpers that capture these values for us). And, if all throws go through some small set of functions that makes it possible to capture those for logging/debugging which is often useful.

Just got back from a trip so ^ is just initial reaction I'll think about this more. Let me know what you think.

**Comment by randombit** [comment]
@ChrisBFX There are some initial improvements in this area in #1744, I would welcome your comments.

**Comment by randombit** [comment]
I've merged #1744 if you can think of something else that would help in this area I would certainly consider it.

**Comment by randombit** [comment]
We have error code enums that are attached onto `Exception`.
